### PR TITLE
Support one-column 2d array in schema enforcement

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

As titled

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pyfunc/test_model_export_with_loader_module_and_data_path.py [hunk 1]
```diff
@@ -928,3 +928,20 @@ def test_log_model_without_specified_conda_env_uses_default_env_with_expected_de
...
+    signature = infer_signature(X, y)
+    assert signature.inputs.inputs[0].shape == (-1, 1)
+    assert signature.outputs.inputs[0].shape == (-1,)
+
+    with mlflow.start_run():
+        model_info = mlflow.sklearn.log_model(model, "model", signature=signature)
+
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    pdf = pd.DataFrame(X)
+    np.testing.assert_almost_equal(loaded_model.predict(pdf), model.predict(pdf))
```
Comment (harupy) [inline tests/pyfunc/test_model_export_with_loader_module_and_data_path.py:947]: On master, this results in the following error:


```
Traceback (most recent call last):
  File "a.py", line 19, in <module>
    np.testing.assert_almost_equal(loaded_model.predict(pdf), model.predict(pdf))
  File "/home/haru/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 415, in predict
    data = _enforce_schema(data, input_schema)
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/utils.py", line 672, in _enforce_schema
    _enforce_tensor_schema(pf_input, input_schema)
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/utils.py", line 570, in _enforce_tensor_schema
    new_pf_input = _reshape_and_cast_pandas_column_values(
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/utils.py", line 470, in _reshape_and_cast_pandas_column_values
    raise MlflowException(
mlflow.exceptions.MlflowException: The input pandas dataframe column 'None' contains scalar values, which requires the shape to be (-1,), but got tensor spec shape of (-1, 1).
```

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -466,14 +466,17 @@ def _reshape_and_cast_pandas_column_values(name, pd_series, tensor_spec):
...
-                error_code=INVALID_PARAMETER_VALUE,
-            )
-        return _enforce_tensor_spec(np.array(pd_series, dtype=tensor_spec.type), tensor_spec)
+        for shape in [(-1,), (-1, 1)]:
+            if tensor_spec.shape == shape:
+                return _enforce_tensor_spec(
+                    np.array(pd_series, dtype=tensor_spec.type).reshape(shape), tensor_spec
+                )
+        raise MlflowException(
+            f"The input pandas dataframe column '{name}' contains scalar "
```
Comment (WeichenXu123) [inline mlflow/models/utils.py:475]: We should also update the error message here if we make this change.

Comment (WeichenXu123) [reply mlflow/models/utils.py:475]:
```suggestion
            f"The input pandas dataframe column '{name}' contains scalar or one dimension array "
```

**Review by WeichenXu123 — APPROVED** [review]
LGMT except one comments.

File: mlflow/models/utils.py (hunk 2 above)
Comment (harupy) [reply mlflow/models/utils.py:475]: @WeichenXu123 We check `np.isscalar(pd_series[0])` on line 468, the column doesn't contain one dimension array.

**Comment by harupy** [comment]
@WeichenXu123 Replied to your comment. See https://github.com/mlflow/mlflow/pull/8253#discussion_r1169671832

File: mlflow/models/utils.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/models/utils.py:475]: oh, yes. you are right.

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**